lib/Parser.js

```diff
@@ -0,0 +1,809 @@
+// @flow
+
+const fs = require('fs-extra');
+
+const LOG_ERRORS = 'ELM_TEST_LOG_PARSE_ERRORS' in process.env;
+
+// For valid Elm files, this extracts _all_ (no more, no less) names that:
+// 1. Are exposed.
+// 2. _Might_ be tests. So capitalized names are excluded, for example.
+//
+// For invalid Elm files, this probably returns an empty list. It could also
+// return a list of things it _thinks_ are exposed values, but it doesn’t
+// matter. The idea is to bail early and still import the file. Then Elm gets a
+// chance to show its nice error messages.
+//
+// The only times the returned promise is rejected are:
+//
+// - When there’s a problem reading the file from disk.
+// - If there’s a bug and an unexpected exception is thrown somewhere.
+// - If `effect module` is encountered. That’s a real edge case. The gist of it
+//   is that applications can’t contain effect modules and since the tests are
+//   compiled as an application we can’t include effect modules.
+//
+// The tokenizer reads the file character by character. As soon as it’s produced
+// a whole token it feeds it to the parser, which works token by token. Both
+// parse just enough to be able to extract all exposed names that could be tests
+// without false positives.
+function extractExposedPossiblyTests(
+  filePath /*: string */,
+  createReadStream /*: typeof fs.createReadStream */ = fs.createReadStream
+) /*: Promise<Array<string>> */ {
+  return new Promise((resolve, reject) => {
+    const exposed /*: Array<string> */ = [];
+
+    let tokenizerState /*: typeof TokenizerState */ = {
+      tag: 'Initial',
+      otherTokenChars: '',
+    };
+
+    let parserState /*: typeof ParserState */ = {
+      tag: 'ModuleDeclaration',
+      lastToken: 'Nothing',
+    };
+
+    let newChunk = true;
+
+    let lastLowerName = '';
+
+    let index = 0;
+
+    const readable = createReadStream(filePath, {
+      encoding: 'utf8',
+      highWaterMark: 4096,
+    });
```

Comment:
This number is completely arbitrary.

According to the [createReadStream docs](https://nodejs.org/api/fs.html#fs_fs_createreadstream_path_options) the default is 64 kB. And the default for the readable stream base class is 16 kB. So 4 kB is way less.

At work 4.0 kB is by far the most common size of all our Elm files (the block size of the file system is 4 kB on macOS):

```
 158 4.0K
  49 8.0K
  20  12K
  14  16K
  13  20K
   6  24K
   5  28K
   2  36K
   4  44K
   1  48K
   1  52K
   2  56K
   1  60K
   1  68K
   1  76K
   1  80K
   1  88K
   1 112K
   1 172K
```

So reading 4 kB at a time reads the entire file most of the time. Which maybe either is fine, or excessive.

I have tried setting it to `1` (which should read 1 char at at time). That seems to be slightly slower. But basically any other number seems to be faster and perform the same.

So maybe it’s all about memory usage. If 4 kB is read at a time per file, you’d need 256 files to use 1 MB of memory. That’s very little, right? At work we have 282 Elm files currently. But way fewer in `tests/`.

At work our average module declaration is 120 characters. Mean 80. Smallest 27, largest 1214. But in `tests/` they only range between 32 and 49.

So it might make sense to have a highWaterMark around that. Unless people usually have `exposing (..)` in their tests. Then it’s maybe more sensible to aim for the average test name length or something. But my guess would be that people expose either everything (`(..)`) or just one or a couple of values per file.

`find-elm-dependencies` (which is used (indirectly) by elm-webpack-loader) uses 60 chars at a time by reading 480 bytes per chunk: https://github.com/NoRedInk/find-elm-dependencies/blob/c355c864e1bd334998d27d750c9fd38558730f39/src/dependencies.js#L8-L9

In conclusion:

- It does not generally affect performance? Maybe when not having an SSD?
- Is it worth chasing the ideal memory usage?
- Can we measure the idea number?

Reply:
So we use this stream to read from an elm file, then we parse the incoming data and produce definitions.

I would not be surprised if we are IO bound here. Certainly I would be surprised if we ever had to ask the OS to stop giving us data because we are processing it too slowly. I think (I may be wrong) that this means that we should never reach the high watermark provided it isn't set to something really small.

Reply:
> So we use this stream to read from an elm file, then we parse the incoming data and produce definitions?

Correct! And unless `exposing (..)` is used, we stop the stream once we’ve parsed the module declaration.

Reply:
I think go with the default value then and trust that it will not matter in practice

Reply:
Switched to default.